File: src/hetzner_code/attach.py
"""Shared logic for wiring one repo onto one already-reachable box —
used by both `create` (box + first repo) and `add` (Nth repo onto an
existing box), so the two never drift apart.
"""

from __future__ import annotations

from pathlib import Path

import click

from hetzner_code import github, keys, ssh_util, state
from hetzner_code.config import REMOTE_CODE_DIR, REMOTE_KEY_DIR
from hetzner_code.naming import generate_key_title
from hetzner_code.provision import clone_repo_command
from hetzner_code.run import HcodeError


def attach_repo(
    *,
    ip: str,
    identity_path: Path,
    instance_name: str,
    repo_url: str,
    branch: str | None,
    env_files: tuple[str, ...],
) -> state.Repo:
    repo = github.parse_repo_url(repo_url)

    click.echo(f"  registering deploy key for {repo.slug} ...")
    key_dir = state.key_dir(instance_name, repo.name)
    private_key, public_key = keys.generate(
        key_dir, comment=f"hcode:{instance_name}:{repo.name}"
    )
    title = generate_key_title(instance_name, repo.name)
    deploy_key_id = github.add_deploy_key(repo, public_key_path=public_key, title=title)

    remote_key_path = f"{REMOTE_KEY_DIR}/{repo.name}"
    dest = f"{REMOTE_CODE_DIR}/{repo.name}"

    try:
        click.echo(f"  cloning {repo.slug} onto the box ...")
        ssh_util.copy_to(ip, private_key, remote_key_path, identity_path)
        ssh_util.run_remote(
            ip,
            clone_repo_command(
                repo_url=repo_url,
                repo_name=repo.name,
                branch=branch,
                key_path=remote_key_path,
                dest=dest,
            ),
            identity_path,
        )

        if env_files:
            click.echo(f"  copying {len(env_files)} env file(s) ...")
            for local in env_files:
                local_path = Path(local)
                if local_path.is_absolute():
                    raise HcodeError(
                        f"--env-file {local} is absolute — pass it relative to the repo "
                        "root (e.g. backend/.env), the same relative path is used on the box"
                    )
                if not local_path.exists():
                    raise HcodeError(f"--env-file {local} does not exist locally")
                remote_path = f"{dest}/{local_path.as_posix()}"
                ssh_util.copy_to(ip, local_path, remote_path, identity_path)
        else:
            click.echo(
                "  no --env-file given — if this repo needs one, copy it up "
                f"yourself or re-run with --env-file (see `hcode status {instance_name}`)"
            )
    finally:
        keys.forget_private_key(private_key)

    return state.Repo(
        url=repo_url,
        owner=repo.owner,
        name=repo.name,
        branch=branch,
        deploy_key_id=deploy_key_id,
        deploy_key_title=title,
    )

File: internal/attach/attach.go
// Package attach holds shared logic for wiring one repo onto one
// already-reachable box — used by both `create` (box + first repo) and
// `add` (Nth repo onto an existing box), so the two never drift apart.
package attach

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/dikee/hetzner-code/internal/config"
	"github.com/dikee/hetzner-code/internal/github"
	"github.com/dikee/hetzner-code/internal/keys"
	"github.com/dikee/hetzner-code/internal/naming"
	"github.com/dikee/hetzner-code/internal/provision"
	"github.com/dikee/hetzner-code/internal/run"
	"github.com/dikee/hetzner-code/internal/sshutil"
	"github.com/dikee/hetzner-code/internal/state"
)

// AttachRepo registers a fresh deploy key for repoURL, clones it onto
// the box, and copies up any envFiles.
func AttachRepo(ip, identityPath, instanceName, repoURL string, branch *string, envFiles []string) (state.Repo, error) {
	repo, err := github.ParseRepoURL(repoURL)
	if err != nil {
		return state.Repo{}, err
	}

	fmt.Printf("  registering deploy key for %s ...\n", repo.Slug())
	keyDir := state.KeyDir(instanceName, repo.Name)
	privateKey, publicKey, err := keys.Generate(keyDir, fmt.Sprintf("hcode:%s:%s", instanceName, repo.Name))
	if err != nil {
		return state.Repo{}, err
	}
	title := naming.GenerateKeyTitle(instanceName, repo.Name)
	deployKeyID, err := github.AddDeployKey(repo, publicKey, title, true)
	if err != nil {
		return state.Repo{}, err
	}

	remoteKeyPath := fmt.Sprintf("%s/%s", config.RemoteKeyDir, repo.Name)
	dest := fmt.Sprintf("%s/%s", config.RemoteCodeDir, repo.Name)

	attachErr := func() error {
		fmt.Printf("  cloning %s onto the box ...\n", repo.Slug())
		if err := sshutil.CopyTo(ip, privateKey, remoteKeyPath, identityPath); err != nil {
			return err
		}
		cloneCmd := provision.CloneRepoCommand(repoURL, repo.Name, branch, remoteKeyPath, dest)
		if _, err := sshutil.RunRemote(ip, cloneCmd, identityPath); err != nil {
			return err
		}

		if len(envFiles) > 0 {
			fmt.Printf("  copying %d env file(s) ...\n", len(envFiles))
			for _, local := range envFiles {
				if filepath.IsAbs(local) {
					return run.Errorf(
						"--env-file %s is absolute — pass it relative to the repo root "+
							"(e.g. backend/.env), the same relative path is used on the box", local,
					)
				}
				if _, err := os.Stat(local); err != nil {
					return run.Errorf("--env-file %s does not exist locally", local)
				}
				remotePath := fmt.Sprintf("%s/%s", dest, filepath.ToSlash(local))
				if err := sshutil.CopyTo(ip, local, remotePath, identityPath); err != nil {
					return err
				}
			}
		} else {
			fmt.Printf(
				"  no --env-file given — if this repo needs one, copy it up yourself or "+
					"re-run with --env-file (see `hcode status %s`)\n", instanceName,
			)
		}
		return nil
	}()

	if err := keys.ForgetPrivateKey(privateKey); err != nil {
		return state.Repo{}, err
	}
	if attachErr != nil {
		return state.Repo{}, attachErr
	}

	return state.Repo{
		URL:            repoURL,
		Owner:          repo.Owner,
		Name:           repo.Name,
		Branch:         branch,
		DeployKeyID:    deployKeyID,
		DeployKeyTitle: title,
		Worktrees:      []string{},
	}, nil
}
